src/corehost/cli/breadcrumbs.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <thread>
+#include "pal.h"
+#include "utils.h"
+#include "trace.h"
+#include "breadcrumbs.h"
+
+static const int SYNCHRONOUS_WRITE_SIZE = 20;
+
+breadcrumb_writer_t::breadcrumb_writer_t(const std::unordered_set<pal::string_t>* files)
+    : m_status(false)
+    , m_files(files)
+{
+    if (!pal::get_default_breadcrumb_store(&m_breadcrumb_store))
+    {
+        m_breadcrumb_store.clear();
+    }
+}
+
+// Begin breadcrumb writing: write synchronously or launch a
+// thread to write breadcrumbs.
+void breadcrumb_writer_t::begin_write()
+{
+    trace::verbose(_X("--- Begin breadcrumb write"));
+    if (m_breadcrumb_store.empty())
+    {
+        trace::verbose(_X("Breadcrumb store was not obtained... skipping write."));
+        m_status = false;
+        return;
+    }
+
+    trace::verbose(_X("Number of breadcrumb files to write is %d"), m_files->size());
+    if (m_files->size() <= SYNCHRONOUS_WRITE_SIZE)
+    {
+        trace::verbose(_X("Writing breadcrumbs on the calling thread"));
+        m_status = write();
+        return;
+    }
+    m_thread = std::thread(thread_callback, this);
```

Comment:
It is already namespaced within `breadcrumb_writer_t`, so `breadcrumb_writer_t::breadcrumb_write_callback` is redundant. I've renamed it to `write_worker_callback`